File: [0200][Medium][Number_of_Islands]/Number_of_Islands_5.py
# leetcode time     cost : 84 ms
# leetcode memory   cost : 13.9 MB 
# Time  Complexity: O(M*N)
# Space Complexity: O(M*N)
# solution 1. DFS
class Solution(object):
    def is_valid(self, grid, r, c):
        m, n = len(grid), len(grid[0])
        if r < 0 or c < 0 or r >= m or c >= n:
            return False
        return True

    def numIslands(self, grid):     #DFS
        if not grid or not grid[0]:
            return 0

        m, n = len(grid), len(grid[0])
        count = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    self.dfs(grid, i, j)
                    count += 1
        return count

    def dfs(self, grid, r, c):
        grid[r][c] = '0'
        directions = [(0,1), (0,-1), (-1,0), (1,0)]
        for d in directions:
            nr, nc = r + d[0], c + d[1]    
            if self.is_valid(grid, nr, nc) and grid[nr][nc] == '1':
                self.dfs(grid, nr, nc)

def main():
    grid = [["1","1","1","1","0"],["1","1","0","1","0"],["1","1","0","0","0"],["0","0","0","0","0"]]  # expect is 1,
    obj = Solution()
    result = obj.numIslands(grid)       
    print("return result is ",result)
    
if __name__ =='__main__':
    main() 

File: [0200][Medium][Number_of_Islands]/Number_of_Islands_5.java
// leetcode time     cost : 3 ms
// leetcode memory   cost : 42.1 MB 
// Time  Complexity: O(M*N)
// Space Complexity: O(M*N)
// solution 2, BFS,
// // https://leetcode.com/problems/number-of-islands/discuss/56349/7-lines-Python-~14-lines-Java
public class Solution {
  public int numIslands(char[][] grid) {
      int islands = 0;
      for (int i=0; i<grid.length; i++)
          for (int j=0; j<grid[i].length; j++)
              islands += sink(grid, i, j);
      return islands;
  }
  int sink(char[][] grid, int i, int j) {
      if (i < 0 || i == grid.length || j < 0 || j == grid[i].length || grid[i][j] == '0')
          return 0;
      grid[i][j] = '0';
      for (int k=0; k<4; k++)
          sink(grid, i+d[k], j+d[k+1]);
      return 1;
  }
  int[] d = {0, 1, 0, -1, 0};
}

public class Number_of_Islands_5 {
    public static void main(String args[]) {
        char[][] grid = {{'1','1','1','1','0'},{'1','1','0','1','0'},{'1','1','0','0','0'},{'0','0','0','0','0'}}; // expect is 1,
        Solution Solution_obj = new Solution();
        int result = Solution_obj.numIslands(grid);
        System.out.println("In java code,result is :" + result);
    }
}
